components/eip681/src/parse.rs

```diff
@@ -0,0 +1,1565 @@
+//! Types and functions used for parsing.
+
+use std::borrow::Cow;
+
+use nom::{AsChar, Parser};
+use snafu::{OptionExt, ResultExt};
+
+use crate::error::*;
+
+/// Zero or more consecutive digits.
+///
+/// ```abnf
+/// *DIGIT
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Digits {
+    places: Vec<u8>,
+}
+
+impl core::fmt::Display for Digits {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for n in self.places.iter() {
+            f.write_fmt(format_args!("{n}"))?;
+        }
+        Ok(())
+    }
+}
+
+impl Digits {
+    #[cfg(any(test, feature = "test-dependencies"))]
+    pub const MAX_PLACES: usize = u64::MAX.ilog10() as usize;
+
+    /// Returns the `u64` representation.
+    ///
+    /// ## Errors
+    /// Errors if internal arithmetic operations overflow.
+    pub fn as_u64(&self) -> Result<u64, ValidationError> {
+        let mut total = 0u64;
+        for (mag, digit) in self.places.iter().rev().enumerate() {
+            let increment = (*digit as u64)
+                .checked_mul(
+                    10u64
+                        .checked_pow(u32::try_from(mag).context(IntegerSnafu)?)
+                        .context(OverflowSnafu)?,
+                )
+                .context(OverflowSnafu)?;
+            total = total.checked_add(increment).context(OverflowSnafu)?;
+        }
+        Ok(total)
+    }
+
+    /// Returns the digits as a ratio, where prefixed zeros
+    /// are treated as a denominator.
```

Comment:
This comment is confusing to me, because there is no period character in `Digits`, so `self` is not a ratio.

I *think* what you are saying is that this method returns the ratio representing the decimal number `0.<digits>`, i.e. re-interpreting `self` as the fractional component of some other decimal number.

Reply:
```suggestion
    /// Returns the ratio corresponding to the decimal number `0.<digits>`,
    /// i.e. the denominator will be `10^len(digits)`.
```